uart_tsi/testchip_uart_tsi.cc

```diff
@@ -142,62 +143,50 @@ bool testchip_uart_tsi_t::check_connection() {
   uint8_t rdata = 0;
   int n = read(ttyfd, &rdata, 1);
   if (n > 0) {
-    printf("Error: Reading unexpected data %c from UART. Abort.\n", rdata);
+    PRINTF("Error: Reading unexpected data %c from UART. Abort.\n", rdata);
     exit(1);
   }
   return true;
 }
 
 void testchip_uart_tsi_t::load_program() {
   in_load_program = true;
+  PRINTF("Loading program\n");
   testchip_tsi_t::load_program();
+  PRINTF("Done loading program\n");
   in_load_program = false;
-
-  uint8_t rbuf[chunk_max_size()];
-  if (do_self_check) {
-    printf("Performing self check\n");
-    for (auto &it : loaded_program) {
-      addr_t addr = it.first;
-      printf("Self check chunk %lx to %lx\n", addr, addr + it.second.size());
-      read_chunk(addr, it.second.size(), rbuf);
-      for (size_t i = 0; i < it.second.size(); i++) {
-	if (rbuf[i] != it.second[i]) {
-	  printf("Self check failed at address %lx %x != %x\n", addr + i, rbuf[i], it.second[i]);
-	  exit(1);
-	}
-      }
-      printf("Self check succeeded chunk %lx to %lx\n", addr, addr + it.second.size());
-    }
-    printf("Self check success\n");
-  }
 }
 
 void testchip_uart_tsi_t::write_chunk(addr_t taddr, size_t nbytes, const void* src) {
+  if (this->in_load_program) { PRINTF("Loading ELF %lx-%lx ... ", taddr, taddr + nbytes); }
   testchip_tsi_t::write_chunk(taddr, nbytes, src);
-  if (in_load_program) {
-    for (auto &it : loaded_program) {
-      addr_t eaddr = taddr + nbytes;
-      if ((taddr >= it.first && taddr  < (it.first + it.second.size())) ||
-	  (eaddr  > it.first && eaddr <= (it.first + it.second.size())) ||
-	  (taddr  < it.first && eaddr  > (it.first + it.second.size()))) {
-	printf("Error: Overlapping sections in loaded program.\n");
-	printf("Write addr: %lx - %lx\n", taddr, eaddr);
-	printf("Conflict addr: %lx - %lx\n", it.first, it.first + it.second.size());
-	exit(1);
+  while (this->handle_uart()) { }
+  if (this->in_load_program) { printf("Done\n"); }
+
+  if (this->do_self_check && this->in_load_program) {
+    uint8_t rbuf[chunk_max_size()];
+    const uint8_t* csrc = (const uint8_t*)src;
+    PRINTF("Performing self check of region %lx-%lx ... ", taddr, taddr + nbytes);
+    read_chunk(taddr, nbytes, rbuf);
```

Comment:
`read_chunk` seems to delegate to `tsi_t::read_chunk` which is a blocking call effectively (https://github.com/riscv-software-src/riscv-isa-sim/blob/master/fesvr/tsi.cc#L52). Why is the `while (handle_uart()) {}` needed below?

Reply:
Ah I think that one is unnecessary. But it also shouldn't hurt.